### DISCUSSION TITLE
Add Cache-Tag header for CDN or possibility to modify the headers dynamically

### DISCUSSION THREAD

**Discussion by maslowivan** [vercel/next.js#54871] [App Router]
Imagine you have the route like `/app/[slug]/page.tsx`.
The [page] can be different - some product like /adidas-boots or category like /shooes. Depends on server data.

1. impossible to set the different headers for those pages using the [headers](https://nextjs.org/docs/pages/api-reference/next-config-js/headers) configuration, because you don't know how the route will look and the header cache tag data should contain something like `product-12345` or `category-12345`.
2. impossible to set the different cache-tag header based on data. All modern CDN cache system based on this header. For example - [Fastly Surrogate-Key](https://docs.fastly.com/en/guides/working-with-surrogate-keys) or [Cloudflare Cache-Tag](https://developers.cloudflare.com/cache/how-to/purge-cache/purge-by-tags/).
3. If I wrap the next.js with the custom server - I can't read the response body, because it's compressed. With compress: false the server simple doesn't work. Here is the [old issue](https://github.com/vercel/next.js/issues/50320).

May be related: https://github.com/vercel/next.js/discussions/52491
May be also related: https://github.com/vercel/next.js/discussions/36089

The easiest solution can be just allow user to set the custom response headers right from the server component. At least we have the function to generate metadata. Maybe we can simple extend it?

**Comment by CobyPear** [comment]
Here are some more related discussions/issues:
https://github.com/vercel/next.js/discussions/51460
https://github.com/vercel/next.js/issues/50914

My interpretation of the implementation of RSC and the new caching heuristics lead me to believe the headers can not be available in an RSC in the same way they were with the pages router and `getServerSideProps`. Hopefully I am wrong and this will be implemented at some point (I would love to try but my inexperience with the next codebase feels insurmountable at the moment), but the lack of response from the Next.js team on any of these issues/discussions doesn't bode well.

**Comment by maslowivan** [comment]
> Here are some more related discussions/issues: #51460 #50914
> 
> My interpretation of the implementation of RSC and the new caching heuristics lead me to believe the headers can not be available in an RSC in the same way they were with the pages router and `getServerSideProps`. Hopefully I am wrong and this will be implemented at some point (I would love to try but my inexperience with the next codebase feels insurmountable at the moment), but the lack of response from the Next.js team on any of these issues/discussions doesn't bode well.

the last one is absolutely the same issue, correct. Haven't found it before

**Comment by maslowivan** [comment]
Temporary workaround here https://github.com/vercel/next.js/issues/50914#issuecomment-1756477185
